dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/DataValueSetController.java

```diff
@@ -120,109 +116,98 @@
-        setNoStore( response );
-
-        OutputStream outputStream = compress( response, attachment, Compression.fromValue( compression ), "xml" );
-
-        adxDataService.writeDataValueSet( adxDataService.getFromUrl( params ), outputStream );
+        getDataValueSet( params, attachment, compression, "xml", response, CONTENT_TYPE_XML_ADX,
+            out -> {
+                try
+                {
+                    adxDataService.writeDataValueSet( adxDataService.getFromUrl( params ), out );
+                }
+                catch ( AdxException ex )
+                {
+                    // this will end up in same exception handler
+                    throw new IllegalStateException( ex.getMessage(), ex );
+                }
+            } );
     }
 
     @GetMapping( produces = CONTENT_TYPE_JSON )
     public void getDataValueSetJson( DataValueSetQueryParams params,
         @RequestParam( required = false ) String attachment,
         @RequestParam( required = false ) String compression,
         HttpServletResponse response )
-        throws IOException
     {
-        response.setContentType( CONTENT_TYPE_JSON );
-        setNoStore( response );
-
-        OutputStream outputStream = compress( response, attachment, Compression.fromValue( compression ), "json" );
-
-        dataValueSetService.writeDataValueSetJson( dataValueSetService.getFromUrl( params ), outputStream );
+        getDataValueSet( params, attachment, compression, "json", response, CONTENT_TYPE_JSON,
+            out -> dataValueSetService.writeDataValueSetJson( dataValueSetService.getFromUrl( params ), out ) );
     }
 
     @GetMapping( produces = CONTENT_TYPE_CSV )
     public void getDataValueSetCsv( DataValueSetQueryParams params,
         @RequestParam( required = false ) String attachment,
         @RequestParam( required = false ) String compression,
-        IdSchemes idSchemes,
         HttpServletResponse response )
-        throws IOException
     {
-        response.setContentType( CONTENT_TYPE_CSV );
-        setNoStore( response );
-
-        OutputStream outputStream = compress( response, attachment, Compression.fromValue( compression ), "csv" );
+        getDataValueSet( params, attachment, compression, "csv", response, CONTENT_TYPE_CSV,
+            out -> dataValueSetService.writeDataValueSetCsv( dataValueSetService.getFromUrl( params ),
+                new PrintWriter( out ) ) );
+    }
 
-        PrintWriter printWriter = new PrintWriter( outputStream );
+    private void getDataValueSet( DataValueSetQueryParams params,
+        String attachment,
+        String compression, String format, HttpServletResponse response, String contentType,
+        Consumer<OutputStream> writeOutput )
+    {
+        response.setContentType( contentType );
+        setNoStore( response );
 
-        dataValueSetService.writeDataValueSetCsv( dataValueSetService.getFromUrl( params ), printWriter );
+        try (
```

Comment:
OBS: The important part here is that we use try-with-resource so the stream is closed in any case. I know that usually the response stream does not need closing explicitly but if we don't do it here the framework will close the inner wrapped "plain" stream which the compression stream wrapped around it will not witness so it never finalizes the archive content of the stream leaving the archive broken.

Reply:
Not with compression. The compressing wrapper cannot know when you are done adding to the stream and the automatic closing will happen on the HTTP output stream wrapped inside so again the compression around it is never closed and thus does not complete the archive.